public/app/features/explore/Explore.tsx

```diff
@@ -314,6 +314,30 @@ export class Explore extends React.PureComponent<Props, ExploreState> {
     );
   }
 
+  renderPanels(width: number) {
+    const {queryResponse, frames} = this.props;
+    const panels: React.ReactNode[] = [];
+
+    for (const key of Object.keys(frames)) {
```

Comment:
What's `frames` prop in this context? Explore.tsx gets pre-aggregated data already split by visualization type (decorators.ts). Maybe `getPanelForVisType` should be in decorators.ts?

Reply:
It's a `{[visType: string]: DataFrame}` structure. As you said that is mapped and grouped in the decorators.ts. I don't think the decorators.ts file should now about the panels and how that is handled. I think those concerns a bit different and would better be separated otherwise the data transform layer would be passing a react component down to the presentation layer to render which seems to me it breaks the separation of concerns.

Reply:
Oh yeah sorry, I meant that the aggregation of data frames per viz type could happen in decorators. The actual component to be used based on viz type still should be picked in the view layer. It's just that I was not sure about `for (const key of Object.keys(frames))` which seems to go over all frames one by one instead of using map of frames aggregated by viz type.

Reply:
Ah yeah probably the naming should be better. The frames is already grouped object but sounds like just an array of frames. Will change that to something better.